src/solve/test.rs

```diff
@@ -2121,3 +2121,110 @@ fn local_and_external_types() {
         goal { forall<T> { IsLocal(Internal<T>) } } yields { "Unique" }
     }
 }
+
+#[test]
+fn fundamental_types() {
+    // NOTE: These tests need to have both Internal and External structs since chalk will attempt
+    // to enumerate all of them.
+
+    // This first test is a sanity check to make sure `Box` isn't a special case.
+    // By testing this, we ensure that adding the #[fundamental] attribute does in fact
+    // change behaviour
+    test! {
+        program {
+            extern struct Box<T> { }
+
+            extern struct External { }
+            struct Internal { }
+        }
+
+        // Without fundamental, Box should never be local
+        goal { forall<T> { not { IsLocal(Box<T>) } } } yields { "Unique" }
+
+        // Without fundamental, both of these are always non-local
+        goal { IsLocal(Box<External>) } yields { "No possible solution" }
+        goal { IsLocal(Box<Internal>) } yields { "No possible solution" }
+    }
+
+    test! {
+        program {
+            #[fundamental]
+            extern struct Box<T> { }
+
+            extern struct External { }
+            struct Internal { }
+        }
+
+        // With fundamental, Box can be local for certain types
+        goal { forall<T> { not { IsLocal(Box<T>) } } } yields { "No possible solution" }
```

Comment:
Just for kicks, let's try to prove:

```
forall<T> { IsLocal(Box<T>) }
```

and

```
not { forall<T> { IsLocal(Box<T>) } }
```

as well. I would expect all of those to fail.